core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/SimplifyFilterPredicate.java

```diff
@@ -0,0 +1,229 @@
+
+        if (expression instanceof SearchedCaseExpression) {
+            SearchedCaseExpression caseExpression = (SearchedCaseExpression) expression;
+            Optional<Expression> defaultValue = caseExpression.getDefaultValue();
+
+            List<Expression> operands = caseExpression.getWhenClauses().stream()
+                    .map(WhenClause::getOperand)
+                    .collect(toImmutableList());
+
+            List<Expression> results = caseExpression.getWhenClauses().stream()
+                    .map(WhenClause::getResult)
+                    .collect(toImmutableList());
+            long trueResultsCount = results.stream()
+                    .filter(result -> result.equals(TRUE_LITERAL))
+                    .count();
+            long notTrueResultsCount = results.stream()
+                    .filter(SimplifyFilterPredicate::isNotTrue)
+                    .count();
+            // all results true
+            if (trueResultsCount == results.size() && defaultValue.isPresent() && defaultValue.get().equals(TRUE_LITERAL)) {
+                return Optional.of(TRUE_LITERAL);
+            }
+            // all results not true
+            if (notTrueResultsCount == results.size() && (defaultValue.isEmpty() || isNotTrue(defaultValue.get()))) {
+                return Optional.of(FALSE_LITERAL);
+            }
+            // one result true, and remaining results not true
+            if (trueResultsCount == 1 && notTrueResultsCount == results.size() - 1 && (defaultValue.isEmpty() || isNotTrue(defaultValue.get()))) {
+                ImmutableList.Builder<Expression> builder = ImmutableList.builder();
+                for (WhenClause whenClause : caseExpression.getWhenClauses()) {
+                    Expression operand = whenClause.getOperand();
+                    Expression result = whenClause.getResult();
+                    if (isNotTrue(result)) {
+                        builder.add(falseOrNullPredicate(operand));
+                    }
+                    else {
+                        builder.add(operand);
+                        return Optional.of(combineConjuncts(metadata, builder.build()));
+                    }
+                }
+            }
+            // all results not true, and default true
+            if (notTrueResultsCount == results.size() && defaultValue.isPresent() && defaultValue.get().equals(TRUE_LITERAL)) {
+                ImmutableList.Builder<Expression> builder = ImmutableList.builder();
+                operands.stream()
+                        .forEach(operand -> builder.add(falseOrNullPredicate(operand)));
+                return Optional.of(combineConjuncts(metadata, builder.build()));
+            }
+
+            // all conditions not true
+            if (operands.stream().allMatch(SimplifyFilterPredicate::isNotTrue)) {
+                return Optional.of(defaultValue.orElse(FALSE_LITERAL));
+            }
+            // initial conditions not true, and next condition true
```

Comment:
It's not immediately obvious what this means or why this is the case here

Reply:
I changed the comment.